src/services/api.ts

```diff
@@ -0,0 +1,205 @@
+import { BigNumber, utils } from "ethers";
+import fetch, { RequestInit } from "node-fetch";
+
+import {
+  normalizeOrder,
+  Order,
+  OrderKind,
+  Signature,
+  SigningScheme,
+  encodeSignatureData,
+} from "../ts";
+
+interface ApiCall {
+  network: string;
+}
+
+interface GetFeeQuery {
+  sellToken: string;
+  buyToken: string;
+  kind: OrderKind;
+  amount: BigNumber;
+}
+interface EstimateTradeAmountQuery {
+  sellToken: string;
+  buyToken: string;
+  kind: OrderKind;
+  amount: BigNumber;
+}
+interface PlaceOrderQuery {
+  order: Order;
+  signature: Signature;
+}
+interface GetExecutedSellAmountQuery {
+  uid: string;
+}
+
+interface OrderDetailResponse {
+  // Other fields are omitted until needed
+  executedSellAmount: string;
+}
+interface GetFeeResponse {
+  amount: string;
+  expirationDate: Date;
+}
+interface EstimateAmountResponse {
+  amount: string;
+  token: string;
+}
+export interface ApiError {
+  errorType: string;
+  description: string;
+}
+interface CallError extends Error {
+  apiError?: ApiError;
+}
+
+function apiKind(kind: OrderKind): string {
+  switch (kind) {
+    case OrderKind.SELL:
+      return "sell";
+    case OrderKind.BUY:
+      return "buy";
+    default:
+      throw new Error(`Unsupported kind ${kind}`);
+  }
+}
+
+function apiSigningScheme(scheme: SigningScheme): string {
+  switch (scheme) {
+    case SigningScheme.EIP712:
+      return "eip712";
+    case SigningScheme.ETHSIGN:
+      return "ethsign";
+    case SigningScheme.EIP1271:
+      return "eip1271";
+    case SigningScheme.PRESIGN:
+      return "presign";
+    default:
+      throw new Error(`Unsupported signing scheme ${scheme}`);
+  }
+}
+
+async function call<T>(
+  route: string,
+  network: string,
+  init?: RequestInit,
+): Promise<T> {
+  const url = `https://protocol-${network}.dev.gnosisdev.com/api/v1/${route}`;
+  const response = await fetch(url, init);
+  const body = await response.text();
+  if (!response.ok) {
+    const error: CallError = new Error(
+      `Calling "${url} ${JSON.stringify(init)} failed with ${
+        response.status
+      }: ${body}`,
+    );
+    try {
+      error.apiError = JSON.parse(body);
+    } catch {
+      // no api error
+    }
+    throw error;
+  }
+  return JSON.parse(body);
+}
+
+export async function getFee({
+  sellToken,
+  buyToken,
+  kind,
+  amount,
+  network,
```

Comment:
nanonit: to make the API more uniform, network should probably be in the same position for all methods (either first or last). Also, since the network doesn't change throughout the "run" of a script it could maybe be a constructor argument for an "API instance" (this way we could also implement a "mockable" version of the API)?

Edit: I see now that we actually have such an instance. In this case should we not export the helper method to force consumers to use the instance rather than the free methods?

Reply:
Without thinking about mocking, I found it a bit annoying to carry around an `api` instance around the function stack, so I preferred to also have standalone methods.
Mocking is a nice idea. How do you think it's best to mock these functions? My idea would be to have a global variable `mocked` set to false in this file and a `onlyMockedCalls()` function that sets this variable to true. Then each call to the functions exported in this PR would check the `mocked` flag before determining its behavior. (This would be another PR.)

Reply:
> network should probably be in the same position for all methods

I wanted to avoid confusion between `sellToken` and `buyToken`, so all parameters are `{`destructured`}` from the input and order should not matter neither in the functions nor in the API object.